src/core/thread/link_metrics.cpp

```diff
@@ -0,0 +1,389 @@
+        memcpy(buf + length, aTypeIdFlags, linkMetricsQueryOptionsTlv.GetLength());
+        length += linkMetricsQueryOptionsTlv.GetLength();
+    }
+
+    // Set Length for Link Metrics Report TLV
+    tlv->SetLength(length - sizeof(Tlv));
+
+    SuccessOrExit(error = Get<Mle::MleRouter>().SendDataRequest(aDestination, tlvs, sizeof(tlvs), 0, buf, length));
+
+exit:
+    return error;
+}
+
+otError LinkMetrics::AppendSingleProbeLinkMetricsReport(Message &            aMessage,
+                                                        uint8_t &            aLength,
+                                                        const otLinkMetrics &aLinkMetrics,
+                                                        const Message &      aRequestMessage)
+{
+    otError                 error = OT_ERROR_NONE;
+    LinkMetricsReportSubTlv metric;
+
+    aLength = 0;
+
+    // Link Metrics Report sub-TLVs
+    if (aLinkMetrics.mPduCount)
+    {
+        metric.Init();
+        metric.SetMetricsTypeId(kTypeIdFlagPdu);
+        metric.SetMetricsValue32(aRequestMessage.GetPsduCount());
+        SuccessOrExit(error = aMessage.Append(&metric, sizeof(Tlv) + metric.GetLength()));
+        aLength += metric.GetSize();
+    }
+
+    if (aLinkMetrics.mLqi)
+    {
+        metric.Init();
+        metric.SetMetricsTypeId(kTypeIdFlagLqi);
+        metric.SetMetricsValue8(aRequestMessage.GetAverageLqi()); // IEEE 802.15.4 LQI is in scale 0-255
+        SuccessOrExit(error = aMessage.Append(&metric, sizeof(Tlv) + metric.GetLength()));
+        aLength += metric.GetSize();
+    }
+
+    if (aLinkMetrics.mLinkMargin)
+    {
+        metric.Init();
+        metric.SetMetricsTypeId(kTypeIdFlagLinkMargin);
+        metric.SetMetricsValue8(
+            LinkQualityInfo::ConvertRssToLinkMargin(Get<Mac::Mac>().GetNoiseFloor(), aRequestMessage.GetAverageRss()) *
+            255 / 130); // Linear scale Link Margin from [0, 130] to [0, 255]
+        SuccessOrExit(error = aMessage.Append(&metric, sizeof(Tlv) + metric.GetLength()));
+        aLength += metric.GetSize();
+    }
+
+    if (aLinkMetrics.mRssi)
+    {
+        metric.Init();
+        metric.SetMetricsTypeId(kTypeIdFlagRssi);
+        metric.SetMetricsValue8((aRequestMessage.GetAverageRss() + 130) * 255 /
+                                130); // Linear scale rss from [-130, 0] to [0, 255]
+        SuccessOrExit(error = aMessage.Append(&metric, sizeof(Tlv) + metric.GetLength()));
+        aLength += metric.GetSize();
+    }
+
+exit:
+    return error;
+}
+
+uint8_t LinkMetrics::GetTypeIdFlagsFromOtLinkMetricsFlags(LinkMetricsTypeIdFlags *aTypeIdFlags,
+                                                          const otLinkMetrics &   aLinkMetricsFlags)
+{
+    uint8_t count = 0;
+    uint8_t typeIdFlagValue;
+
+    if (aLinkMetricsFlags.mPduCount)
+    {
+        typeIdFlagValue = kTypeIdFlagPdu;
+        memcpy(&aTypeIdFlags[count++], &typeIdFlagValue, sizeof(uint8_t));
```

Comment:
Simplify here and below?

```suggestion
        aTypeIdFlags[count++] = kTypeIdFlagPdu;
```

Reply:
I added a `SetRawValue` method for `LinkMetricsTypeIdFlags` because here assignment operator cannot directly be used(setting `LinkMetricsTypeIdFlags` to an enum).